sdks/python/apache_beam/runners/worker/opcounters.py

```diff
@@ -24,11 +24,118 @@
 import math
 import random
 
+from apache_beam.utils import counters
 from apache_beam.utils.counters import Counter
+from apache_beam.utils.counters import CounterName
 
 # This module is experimental. No backwards-compatibility guarantees.
 
 
+class TransformIOCounter(object):
+  """Class to track time and bytes consumed while reading from IO.
+
+  Subclasses should be able to track consumption of IO across steps
+  in the same stage - for instance, if a Shuffle or Side Input iterable
+  is passed down to a next step.
+
+  Some examples of IO can be side inputs, shuffle, or streaming state.
+  """
+
+  def check_step(self):
+    """Check the current step within a stage as it may have changed.
+
+    If the state changed, it would mean that an initial step passed a
+    data-accessor (such as a side input / shuffle Iterable) down to the
+    next step in a stage.
+    """
+    pass
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    """Exit the IO state."""
+    pass
+
+  def __enter__(self):
+    """Enter the IO state. This should track time spent blocked on IO."""
+    pass
+
+  def add_bytes_read(self, count):
+    pass
+
+
+class SideInputReadCounter(TransformIOCounter):
+  """Tracks time and bytes consumed while reading from side inputs.
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
```

Comment:
<!--thread_id:cc_163113317_t; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:1-->
<!--section:context-quote-->
> **pabloem** wrote:
> It is not very well detailed, it seems. I can detail it further.
> https://github.com/apache/beam/blob/master/sdks/python/apache_beam/utils/counters.py#L31

<!--section:body-->
I've added documentation for io target to counters.py. Check it out.